### DISCUSSION TITLE
Bundle `pip` with Embeddable Python?

### DISCUSSION THREAD

**Discussion by pc223** [Flow-Launcher/Flow.Launcher#644] [Ideas]
**Why:**
For users want to start develop Python plugins without have to install another Python to the system.

If User does not develop Python plugins, then it doesn't affect the current design in any way (packages management is still the same strategy, still `pip install --target`)

**How:**
We need to maintain a workflow like below, this workflow download the official EmbeddablePython, extract, install `pip` (and `setuptools`, `wheel`) into it (with `get-pip.py`), then re-archive it and release as a zip file.

Workflow: https://github.com/pc223/PythonEmbeddableWithPip/blob/main/.github/workflows/test2.yml
Release example: https://github.com/pc223/PythonEmbeddableWithPip/releases

After that we need to add the new .zip url to Droplex for Flow download (maybe automate with CI too)

Thought? 😄

**Comment by jjw24** [comment]
hmm, the concept of embeddable is to include a light weight interpreter with your application, thats why pip is left out, and I think this should also hold true for what comes with flow's python implementation, it is geared towards average user. We can update the python template to include these tools for development, but i dont think pip and dependent packages should be included with the download. Python devs would have their own setup anyway.

Ideally, the development is local and the plugin release and flow's python are both purely for usage, no included extras.

**Reply by pc223** [reply]
This will help beginners get in plugins development, besides, it costs us and users nearly nothing to maintain this.

The new zip file is only **+5MB** more than normal, the CI script above handle everything, and we don't even need to run it that often, maybe 1-2 times per **year** 😄

One reason I think Javascript is so popular because people can easily have access to basic dev env, open browser, press F12, done. I want the same experience for Flow.

Another point, a lot Flow's users are tinkerers and hobbyists, it's likely they want to messing with the plugins, separate dev and prod here does really solve many problems, only restrict the growth of plugins.

**Reply by taooceros** [reply]
I don't think it is easy for beginners to use embedded python to develop. They needs to know where flow installs the python, copy the path to run the code (or set environment variables themselves, which is even more messy that even I hate to do so).
If they want to develop python plugin, just download the normal python installer will be much easier (or download it from Microsoft store).

**Reply by jjw24** [reply]
> I don't think it is easy for beginners to use embedded python to develop. They needs to know where flow installs the python, copy the path to run the code (or set environment variables themselves, which is even more messy that even I hate to do so).
> If they want to develop python plugin, just download the normal python installer will be much easier (or download it from Microsoft store).

This is a good point. what i find developing with embedded is that i have to reference the path for both python and pip if i dont want to mess with env variables, Windows' env variables just doesnt work as easy as Linux's, and i doubt beginners will have an easy time doing this, as oppose to just installing it from the store/website the full package and just works out of the box.

**Reply by deefrawley** [reply]
It could also cause a headache for plugin developers if people are modifying our plugins, then open issues with us to help fix. While plugins are great to extend Flow, I think the entry level should be that you know how to setup a python dev environment on your own PC. Otherwise people can request plugins that do certain things via Discord or Discussions.

**Comment by jjw24** [comment]
what about a plugin that will download the python template and embedded python with pip that users can use to tinker/develop plugins? this saves from copy and pasting code from the python doco guide, plus the benefit of a dedicated plugin dev environment?

**Reply by pc223** [reply]
I think it's too much complexity to avoid 5MB.

Right now without `pip`, users already can use `notepad` to modify Python plugins code directly, and see the effect immediately, without restart Flow. That's the benefit of having Python. The development cycle is shorten a lot. No special compiler/SDK need to be downloaded.

With `pip` bundled out of the box, users can download (pip install --target) additional libraries for existed plugins or new plugins. So the only thing users need to develop Python plugins is `notepad` :smile:

**Reply by jjw24** [reply]
Hmm, @Flow-Launcher/team thoughts?

**Reply by taooceros** [reply]
> I don't think it is easy for beginners to use embedded python to develop. They needs to know where flow installs the python, copy the path to run the code (or set environment variables themselves, which is even more messy that even I hate to do so).
> If they want to develop python plugin, just download the normal python installer will be much easier (or download it from Microsoft store).

**Reply by Garulf** [reply]
I believe having more options is better, but-- I feel embedded Python should remain as static as possible. If the user wants to make modifications they should do as @taooceros mentions above.

> With pip bundled out of the box, users can download (pip install --target) additional libraries for existed plugins or new plugins. So the only thing users need to develop Python plugins is notepad 😄

Can you please provide a scenario where a Plugin would need additional libraries?

**Reply by jjw24** [reply]
i think @pc223 means that users may modify the plugin on the fly and may pull in additional packages for those modifications.

**Reply by Garulf** [reply]
> i think @pc223 means that users may modify the plugin on the fly and may pull in additional packages for those modifications. 

Ah, okay I don't fully understand but I see no downside other then adding a bigger download.

**Reply by pc223** [reply]
> i think @pc223 means that users may modify the plugin on the fly and may pull in additional packages for those modifications.

Yes, this is what I meant 👍 😄

**Reply by taooceros** [reply]
Understand. Though, it is still not that easy to use the embedded python. If they want to do so, install a fresh python may be better. They can install the dependency to the lib folder in plugin I think?

**Reply by taooceros** [reply]
However, if it is easy to maintain, I think include pip may not be troublesome, but just I don't think a lot of people may use it.

**Reply by pc223** [reply]
> However, if it is easy to maintain,

I think it's pretty easy to maintain, just point the download link to a repo like this and we done, maybe make 1 or 2 releases per **year** is enough :smile:

https://github.com/pc223/PythonEmbeddableWithPip/releases/

**Comment by Zeroto521** [comment]
I don't suggest do this.

Hi, don't forget why we use the pure python interpreter.
For normal users could install the python plugin and use it quickly.
This way normal users could avoid download, install, and key in the path of interpreter those steps.

But for advanced users, if they want to develop a plugin.
The above things wouldn't get problems for them.
The only embedded python even with pip is not a **good choice**.
And there are a lot of good-packed interpreters. Help to handle 3rd package installing source problem, compatibility problem, and more.
If they use pure interpreter would get more troblems.

**Comment by pc223** [comment]
@taooceros @jjw24 @Garulf @deefrawley 
I don't think it's that hard to work with embed-py even for beginner, they don't have to deal with `PATH` if they don't want to, @jjw24  just `.\PythonEmbeddable\python.exe -m pip install ...`, you don't have to config any thing about `PATH` to use `pip` (with the setup I propose)

Right now, without `pip`, beginner already can open the plugins in `AppData` and modified it directly, no need to install and additional compiler/SDK, with just `notepad` and you can start hacking right away.

With `pip` install like proposed in this topic, if users need to install external lib, for example `requests`, the command gonna be like this:

```
$ cd C:\Users\user1\AppData\Roaming\FlowLauncher\
$ .\PythonEmbeddable\python.exe -m pip install requests --target .\Plugins\Currency Converter-v1.2.2\lib
```

Not that complicated, we can write this in a beginner tutorial for people copy, seems much simpler than install the offical python, which can cause troubles in multiple ways: 
- maybe they just lazy :smile:
- maybe they have trouble with `PATH`, `PYTHONPATH`,... (dozen of StackOverflow questions about this)
- maybe they get confuse by 2 `python` in 1 system
- maybe they already have python in the system and don't know it
- maybe mismatch between the python version of the system and the embed-py (python27 vs python38)
- maybe mismatch between the pip version of the system and the embed-py

Too many "maybe" for beginners, with `pip` already bundled in, which practically cost us nothing, users can just download Flow, and start hacking right away.
